java-squid/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java

```diff
@@ -410,6 +411,29 @@ private void visit(Tree tree, Tree terminator) {
     if (tree.parent().is(Tree.Kind.EXPRESSION_STATEMENT)) {
       programState = ProgramState.unstack(programState, programState.stack.size()).a;
     }
+    removeUnreachableSymbolicValues();
+  }
+
+  private void removeUnreachableSymbolicValues() {
+    final List<SymbolicValue> unreachableValues = new ArrayList<>();
+    programState.constraints.forEach(new PMap.Consumer<SymbolicValue, Object>() {
+      @Override
+      public void accept(SymbolicValue sv, Object value) {
+        if(sv instanceof SymbolicValue.ObjectSymbolicValue && !programState.isReachable(sv)) {
+          unreachableValues.add(sv);
+        }
+      }
+    });
+
+    PMap<SymbolicValue, Object> cleanedConstraints = programState.constraints;
+    for (SymbolicValue unreachableValue : unreachableValues) {
+      cleanedConstraints = cleanedConstraints.remove(unreachableValue);
+    }
+
+    if(programState.constraints != cleanedConstraints) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Change this comparison to use the equals method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1698)